osier/models/dispatch.py

```diff
@@ -0,0 +1,210 @@
+        return dict(zip(self.tech_set, capacity_set))
+
+    @property
+    def time_set(self):
+        return range(self.n_hours)
+
+    @property
+    def indices(self):
+        return list(it.product(self.tech_set, self.time_set))
+
+    @property
+    def cost_params(self):
+        v_costs = np.array([
+            (t.variable_cost_ts(self.n_hours))
+            for t in self.technology_list
+        ]).flatten()
+        return dict(zip(self.indices, v_costs))
+
+    @property
+    def upper_bound(self):
+        caps = unyt_array([t.capacity for t in self.technology_list])
+        return caps.max().to_value()
+
+    def _create_model_indices(self):
+        self.model.U = pe.Set(initialize=self.tech_set, ordered=True)
+        self.model.T = pe.Set(initialize=self.time_set, ordered=True)
+
+    def _create_demand_param(self):
+        self.model.D = pe.Param(self.model.T, initialize=dict(
+            zip(self.model.T, self.net_demand)))
+
+    def _create_cost_param(self):
+        self.model.C = pe.Param(
+            self.model.U,
+            self.model.T,
+            initialize=self.cost_params)
+
+    def _create_model_variables(self):
+        self.model.x = pe.Var(self.model.U, self.model.T,
+                              domain=pe.Reals,
+                              bounds=(self.lower_bound, self.upper_bound))
+
+    def _objective_function(self):
+        expr = sum(self.model.C[u, t] * self.model.x[u, t]
+                   for u in self.model.U for t in self.model.T)
+        self.model.objective = pe.Objective(sense=pe.minimize, expr=expr)
+
+    def _oversupply_constraint(self):
+        self.model.oversupply = pe.ConstraintList()
+        for t in self.model.T:
+            generation = sum(self.model.x[u, t] for u in self.model.U)
+            over_demand = self.model.D[t] * (1 + self.oversupply)
+            self.model.oversupply.add(generation <= over_demand)
+
+    def _undersupply_constraint(self):
+        self.model.undersupply = pe.ConstraintList()
+        for t in self.model.T:
+            generation = sum(self.model.x[u, t] for u in self.model.U)
+            under_demand = self.model.D[t] * (1 - self.undersupply)
+            self.model.undersupply.add(generation >= under_demand)
+
+    def _generation_constraint(self):
+        self.model.gen_limit = pe.ConstraintList()
+        for t in self.model.T:
+            for u in self.model.U:
+                unit_generation = self.model.x[u, t]
+                unit_capacity = self.capacity_dict[u].to_value()
+                self.model.gen_limit.add(unit_generation <= unit_capacity)
+
+    def _write_model_equations(self):
+
+        self._create_model_indices()
+        self._create_demand_param()
+        self._create_cost_param()
+        self._create_model_variables()
+        self._objective_function()
+        self._oversupply_constraint()
+        self._undersupply_constraint()
+        self._generation_constraint()
+
+        return
```

Comment:
I did this because I don't want the class to initialize the equations until it's time to solve the problem. That way you can update and change variables more quickly. I wrapped them inside a second helper function just to make the code a bit more readable, IMO.